File: interview_code/11_min_number_in_rotated_array/c/min_number_in_rotated_array.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


int min_number_in_rotated_array(int *numbers, int count, int *return_number)
{
	int i;
	int min;
	
	if(NULL == numbers || count <= 0){
		return -1;
	}
	min = numbers[0];
	for(i = 1; i < count; i++){
		if(numbers[i] < min){
			min = numbers[i];
		}
	}
	*return_number = min;
	return 0;
}


int get_min(int *numbers, int count, int *return_number)
{
	if(NULL == numbers || count <= 0){
		return -1;
	}
	
	int start_index = 0;
	int end_index = count-1;
	int middle_index = start_index;
	while(numbers[start_index] >= numbers[end_index]){
		if(end_index - start_index == 1){
			middle_index = end_index;
			break;
		}
		
		middle_index = (start_index+end_index)/2;
		
		if(numbers[start_index] == numbers[middle_index] && numbers[middle_index] == numbers[end_index]){
			return min_number_in_rotated_array(numbers, count, return_number);
		}
		if(numbers[middle_index] >= numbers[end_index]){
			start_index = middle_index;
		}else{
			end_index = middle_index;
		}
	}
	*return_number = numbers[middle_index];
	return 0;
}

// ====================测试代码====================
void test(int* numbers, int length, int expected_1, int expected_2)
{
    int result = 0;
	int i;
	
	for(i = 0; i < length; ++i){
		printf("%d ", numbers[i]);
	}
	
	if(expected_2 == get_min(numbers, length, &result)){
		if(result == expected_1){
			printf("\tpassed\n");
		}else{
			printf("\tfailed\n");
		}
	}else{
		printf("test failed.\n");
	}
}

int main(int argc, char* argv[])
{
    // 典型输入，单调升序的数组的一个旋转
    int array1[] = { 3, 4, 5, 1, 2 };
    test(array1, sizeof(array1) / sizeof(int), 1, 0);

    // 有重复数字，并且重复的数字刚好的最小的数字
    int array2[] = { 3, 4, 5, 1, 1, 2 };
    test(array2, sizeof(array2) / sizeof(int), 1, 0);

    // 有重复数字，但重复的数字不是第一个数字和最后一个数字
    int array3[] = { 3, 4, 5, 1, 2, 2 };
    test(array3, sizeof(array3) / sizeof(int), 1, 0);

    // 有重复的数字，并且重复的数字刚好是第一个数字和最后一个数字
    int array4[] = { 1, 0, 1, 1, 1 };
    test(array4, sizeof(array4) / sizeof(int), 0, 0);

    // 单调升序数组，旋转0个元素，也就是单调升序数组本身
    int array5[] = { 1, 2, 3, 4, 5 };
    test(array5, sizeof(array5) / sizeof(int), 1, 0);

    // 数组中只有一个数字
    int array6[] = { 2 };
    test(array6, sizeof(array6) / sizeof(int), 2, 0);

    // 输入NULL
    test(NULL, 0, 0, -1);

    return 0;
}

File: interview_code/11_min_number_in_rotated_array/python/min_number_in_rotated_array.py
# -*- coding: utf-8 -*-

def get_min_each(numbers, start_index, end_index):
    if numbers == None or end_index >= len(numbers) or start_index < 0 or start_index > end_index:
        raise('argument error!')
        
    min_number = numbers[end_index]
    for i in numbers[start_index:end_index]:
        if i < min_number:
            min_number = i
            
    return min_number

def get_min(numbers):
    if None == numbers or len(numbers) <= 0:
        raise('argument error!')
    
    start_index = 0
    end_index = len(numbers) - 1
    middle_index = start_index
    
    while numbers[start_index] >= numbers[end_index]:
        if end_index - start_index == 1:
            middle_index = end_index
            break
        middle_index = (start_index+end_index)//2
        
        if numbers[start_index] == numbers[middle_index] and numbers[middle_index] == numbers[end_index]:
            return get_min_each(numbers, start_index, end_index)
        
        if numbers[middle_index] >= numbers[start_index]:
            start_index = middle_index
        elif numbers[middle_index] <= numbers[end_index]:
            end_index = middle_index
            
    return numbers[middle_index]



def test(numbers, expetced):
    try:
        min_number = get_min(numbers)
        for i in numbers:
            print(i, end=' ')
            
        if min_number == expetced:
            print('test successed\n')
        else:
            print('test failed\n')
    except:
        if expetced == None:
            print('test successed\n')
        else:
            print('test failed\n')
            
            
            
if __name__ == '__main__':
    array1 = [3, 4, 5, 1, 2]
    test(array1, 1)
    
    array2 = [3, 4, 5, 1, 1, 2]
    test(array2, 1)

    array3 = [3, 4, 5, 1, 2, 2]
    test(array3, 1)

    array4 = [1, 0, 1, 1, 1]
    test(array4, 0)

    array5 = [1, 2, 3, 4, 5]
    test(array5, 1)

    array6 = [2]
    test(array6, 2)

    test(None, None)
            
            
        